File: Algorithms/142. Linked List Cycle II/golang/solution.hash-table.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func detectCycle(head *ListNode) *ListNode {
	var visited = make(map[*ListNode]bool)

	node := head
	for node != nil {
		if visited[node] {
			return node
		}
		visited[node] = true
		node = node.Next
	}
	return nil
}

File: Algorithms/142. Linked List Cycle II/kotlin/solution.hash-table.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */

class Solution {
    fun detectCycle(head: ListNode?): ListNode? {
        val visited = mutableSetOf<ListNode>()

        var node = head
        while (node != null) {
            if (visited.contains(node)) {
                return node
            }
            visited.add(node)
            node = node.next
        }
        return null
    }
}